### PR TITLE
ddl: implement the check constraints in create table

### PR DESCRIPTION
**Pull request by AilinKid**
Signed-off-by: Arenatlx <314806019@qq.com>

### What problem does this PR solve?

Close issue: https://github.com/pingcap/tidb/issues/17406
Problem Summary:
Implement check constraints in TiDB
For example:
```
create table t(a int constraint haha check(a>1), constraint hehe check(a<10));
```
the column-level `haha` check constraint and table-level `hehe` constraint now can be checked and attach-stored in tableInfo.

### What is changed and how it works?

What's Changed:
1: modify tableInfo add constraint field.
2: handle the constraint from parser, do dependency check in ddl, and store it.
3: do some test about it

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Breaking backward compatibility

### Release note 
- implement the check constraints in create table

### CODE REVIEW

**Comment by AilinKid** [comment]
run-unit-test

**Comment by AilinKid** [comment]
PTAL @imtbkcat

File: ddl/constraint_test.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	c.Assert(constrs[1].Table.L, Equals, "t")
+	c.Assert(constrs[1].State, Equals, model.StatePublic)
+	c.Assert(len(constrs[1].ConstraintCols), Equals, 1)
+	c.Assert(constrs[1].ConstraintCols[0], Equals, model.NewCIStr("a"))
+	c.Assert(constrs[1].Name, Equals, model.NewCIStr("t_chk_1"))
+	c.Assert(constrs[1].ExprString, Equals, "`a` > 1")
+
+	// Test column-type check constraint fail on dependency.
+	s.tk.MustExec("drop table t")
+	_, err := s.tk.Exec("create table t(a int not null check(b>0))")
```
Comment (imtbkcat) [inline ddl/constraint_test.go:104]: please add a test like `create table t(a int not null check(a > '12345'))`

Comment (AilinKid) [reply ddl/constraint_test.go:104]: nice catch

File: table/tables/tables.go [hunk 2]
```diff
@@ -51,17 +51,19 @@ import (
...
-	sequence        *sequenceCommon
+	physicalTableID   int64
+	Columns           []*table.Column
+	PublicColumns     []*table.Column
+	VisibleColumns    []*table.Column
+	HiddenColumns     []*table.Column
+	WritableColumns   []*table.Column
+	writableIndices   []table.Index
+	indices           []table.Index
+	PublicConstraints []table.Constraint
```
Comment (imtbkcat) [inline table/tables/tables.go:62]: what's the difference between `PublicConstraints` and  `Constraints`

Comment (AilinKid) [reply table/tables/tables.go:62]: maybe should be renamed as `writableConstraints`
the newly added constraint will stay in a state like `write_only`, in this state the constraint will fetch table rows and check it, meantime, the business insert should also follow this constraint but it still not completed.

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -1051,6 +1063,26 @@ func checkDuplicateConstraint(namesMap map[string]bool, name string, foreign boo
...
+			for {
+				// loop until find constrName that haven't been used.
+				if !namesMap[constrName] {
+					namesMap[constrName] = true
+					break
+				}
+				cnt++
+				constrName = fmt.Sprintf("%s%d", constraintPrefix, cnt)
+			}
+			constr.Name = constrName
```
Comment (imtbkcat) [inline ddl/ddl_api.go:1081]: move `namesMap[constrName] = true` to here is better. there may be bug if some one name its check `tableName_chk_1`, `tableName_chk_2`

Comment (AilinKid) [reply ddl/ddl_api.go:1081]: I think here is ok~
the passed parameter `namesMap` has record the used defined name already.

if some one name its check `tableName_chk_1`, `tableName_chk_2`, the loop will run for 3 time, and find `namesMap[3-times] = false` and break

**Comment by imtbkcat** [comment]
ping @AilinKid

**Comment by AilinKid** [comment]
> ping @AilinKid

wait for a while, it can't be merge into master, there is a special branch...

**Comment by AilinKid** [comment]
this PR has migrated to https://github.com/bb7133/tidb/pull/9

**Pull request closed without merging**